README.md
# Bigcommerce for Node.js

[![Build Status](https://travis-ci.org/getconversio/node-bigcommerce.svg?branch=master)](https://travis-ci.org/getconversio/node-bigcommerce)

A node module for authentication and use with the BigCommerce API

## Installation

To install the module using NPM:

```
npm install node-bigcommerce
```

Or Yarn:
```
yarn add node-bigcommerce
```

## Setup

Include the 'node-bigcommerce' module within your script and instantiate it with a config:

```javascript
const BigCommerce = require('node-bigcommerce');

const bigCommerce = new BigCommerce({
  logLevel: 'info',
  clientId: '128ecf542a35ac5270a87dc740918404',
  secret: 'acbd18db4cc2f85cedef654fccc4a4d8',
  callback: 'https://myapplication.com/auth',
  responseType: 'json',
  apiVersion: 'v3' // Default is v2
});
```

##### Instantiating a BigCommerce instance without a config object will result in an error

## Authorization

Set up your Big Commerce as above and pass the following configuration options in:

```
{
  clientId: 'Your application's client ID',
  secret: 'Your secret',
  callback: 'The location you want the app to return to on success',
  responseType: 'json'
}
```

You will be able to get your Client ID and Secret within your application setup. Below is an example using Express' routes:

```javascript
const express = require('express'),
  router = express.Router(),
  BigCommerce = require('node-bigcommerce');

const bigCommerce = new BigCommerce({
  clientId: '128ecf542a35ac5270a87dc740918404',
  secret: 'acbd18db4cc2f85cedef654fccc4a4d8',
  callback: 'https://myapplication.com/auth',
  responseType: 'json'
});

router.get('/auth', (req, res, next) => {
  bigCommerce.authorize(req.query)
    .then(data => res.render('integrations/auth', { title: 'Authorized!', data: data })
    .catch(next);
  });
});
```

The `authorize` method requires the query parameters from the request to be passed. These are required to request a permanent access token which will be passed back in the data object.

An example data object:

```
{
  access_token: '9df3b01c60df20d13843841ff0d4482c',
  scope: 'store_v2_orders_read_only store_v2_products_read_only users_basic_information store_v2_default',
  user: {
    id: 12345,
    username: 'John Smith',
    email: 'john@success.com'
  },
  context: 'stores/x43tqo'
}
```

From this object you can store the `access_token` for re-use when calling the Big Commerce API.

## Load & Uninstall

The only configuration element required to use the `verify` method (used for both load and uninstall endpoints) is `secret`. Below is an example using Express' routes:

```javascript
const express = require('express'),
  router = express.Router(),
  BigCommerce = require('node-bigcommerce');

const bigCommerce = new BigCommerce({
  secret: 'acbd18db4cc2f85cedef654fccc4a4d8',
  responseType: 'json'
});

router.get('/load', (req, res, next) => {
  try {
    const data = bigCommerce.verify(req.query['signed_payload']);
    res.render('integrations/welcome', { title: 'Welcome!', data: data });
  } catch (err) {
    next(err);
  }
});
```

The `verify` method requires the `signed_payload` query parameter to be passed from the request. This is used to verify that the request has come from Big Commerce. The `verify` method returns the following object:

```
{
  user: {
    id: 12345,
    email: 'john@success.com'
  },
  context: 'stores/x43tqo',
  store_hash: 'x43tqo',
  timestamp: 1421748597.4395974
}
```

This will allow you to automatically log the user in (if required) when BigCommerce calls the load endpoint or remove/label a user that has uninstalled your application from their Big Commerce account.

## Calling the API

The API can be called once the user has been authorized and has an access token. There is a helper for each type of request available within Big Commerce (GET, POST, PUT, DELETE).

To make an API Request you will need the following minimum configuration:

```
{
  clientId: 'Your application's client ID',
  accessToken: 'Token assigned to the user during authorisation',
  storeHash: 'The short hash for the store',
  responseType: 'json'
}
```
Parameters that are added to the url need to be escaped before they are passed as part of the path of any call:

```javascript
const path = '/products?name=' + escape('Plain T-Shirt');
```


### GET

The `Get` call requires a path: get(path):

```javascript
const BigCommerce = require('node-bigcommerce');

const bigCommerce = new BigCommerce({
  clientId: '128ecf542a35ac5270a87dc740918404'
  accessToken: '9df3b01c60df20d13843841ff0d4482c',
  responseType: 'json'
});

bigCommerce.get('/products')
  .then(data => {
    // Catch any errors, or handle the data returned
  });
```

### POST & PUT

The 'POST' & 'PUT' calls requires a path with optional data to be sent: post(path, data):

```javascript
var BigCommerce = require('node-bigcommerce');

var bigCommerce = new BigCommerce({
  clientId: '128ecf542a35ac5270a87dc740918404'
  accessToken: '9df3b01c60df20d13843841ff0d4482c',
  responseType: 'json'
});

var product = {
  name: 'Plain T-Shirt',
  type: 'physical',
  description: 'This timeless fashion staple will never go out of style!',
  price: '29.99',
  categories: [18],
  availability: 'available',
  weight: '0.5'
}

// Replace 'post' with 'put' for a put call
bigCommerce.post('/products', product)
  .then(data => {
  // Catch any errors, or handle the data returned
  });
```

### DELETE

The 'DELETE' call requires a path: delete(path). A delete call will not return any data and will return a response status of 204.

```javascript
const BigCommerce = require('node-bigcommerce');

const bigCommerce = new BigCommerce({
  clientId: '128ecf542a35ac5270a87dc740918404',
  accessToken: '9df3b01c60df20d13843841ff0d4482c'
});

bigCommerce.delete('/products/' + productId)
  .then(() => {
  // Catch any errors, data will be null
  });
```

## Debugging

We use `debug`, so just run with environment variable DEBUG set to `node-bigcommerce:*`

```js
$ DEBUG=node-bigcommerce:* node my_test.js
```

## Response Type

You may require the Big Commerce API to return data in a specific format. To return in either JSON or XML just add a 'responseType' to the config:

```javascript
const BigCommerce = require('node-bigcommerce');

const bigCommerce = new BigCommerce({
  logLevel: 'info',
  clientId: '128ecf542a35ac5270a87dc740918404',
  accessToken: '9df3b01c60df20d13843841ff0d4482c',
  responseType: 'xml'
});

bigCommerce.post('/products?name=' + escape('Plain T-Shirt'))
  .then(data => {
  // Catch any errors, data will be null
  });
```

Note that when returning in JSON the data will be parsed into an object, XML will not, and will return a string. When no response type is given the type will resort to whatever the BigCommerce default is.

Webhooks can only be JSON so when dealing with the '/hooks' endpoint leave the responseType blank (or null).

## Testing

```
yarn test
```

## Contributing

This module was originally written to be used with [Conversio](https://conversio.com) and is used in a production environment currently. This will ensure that this module is well maintained, bug free and as up to date as possible.

Conversio's developers will continue to make updates as often as required to have a consistently bug free platform, but we are happy to review any feature requests or issues and are accepting constructive pull requests.


CHANGELOG.md
4.0.0
=====
- [breaking] Updated to `async/await`
- [breaking] Removed body from error text, added as property in error (#62)
- [breaking] Added gzip support (#60)

3.0.0
=====

 - Removed all callbacks, only Promise is supported now.
 - Removed `authorise` method in favor of `authorize`.
 - Removed `callback` method, `verify` will now throw an Error.
 - Removed the logger, use `DEBUG=node-bigcommerce:*` as an environment variable to get debug messages.
 - Dropped support of node-v4 for missing `class` implementation.
 - Removed [catalogue](https://github.com/getconversio/node-bigcommerce/pull/18) hack, please use `apiVersion` configuration property.


test/logger.js
'use strict';

var Request = require('../lib/request'),
  logger = require('../lib/logger'),
  should = require('chai').should(),
  sinon = require('sinon');

describe('Logger', function() {
  it('should be set to a log level of zero by default', function() {
    logger.level.should.equal(0);
  });

  it('Should log an error at any log level', function() {
    var stub = sinon.stub(console, 'log');

    logger.error('Logging test error at 0');
    logger.level = 1;
    logger.error('Logging test error at 1');

    sinon.assert.calledTwice(stub);
    stub.restore();
  });

  it('should log info at level one', function() {
    var stub = sinon.stub(console, 'log');

    logger.level = 1;
    logger.info('Logging test error at 1');

    sinon.assert.calledOnce(stub);
    stub.restore();
  });

  it('should not log info at level zero', function() {
    var stub = sinon.stub(console, 'log');

    logger.level = 0;
    logger.info('Logging test error at 1');

    sinon.assert.notCalled(stub);
    stub.restore();
  });
});


test/helpers.js
'use strict';

var reload = require('require-reload')(require),
  helpers = require('../lib/helpers'),
  logger = require('../lib/logger'),
  should = require('chai').should(),
  sinon = require('sinon');

describe('Helpers', function() {
  describe('defer', function() {
    it('Should create a deferred promise that resolves', function() {
      var deferred = helpers.defer();
      deferred.resolve(123);
      return deferred.promise
        .then(function(result) {
          result.should.equal(123);
        });
    });

    it('Should create a deferred promise that rejects', function(done) {
      var deferred = helpers.defer();
      deferred.reject(new Error('oh no'));
      deferred.promise
        .then(function() {
          done(new Error('Should not resolve'));
        })
        .catch(function(err) {
          err.message.should.equal('oh no');
          done();
        });
    });

    it('should work when there is no global promise', function() {
      var oldPromise = global.Promise;
      global.Promise = undefined;

      helpers = reload('../lib/helpers');

      var deferred = helpers.defer();
      deferred.resolve(123);
      return deferred.promise
        .then(function(result) {
          result.should.equal(123);
          global.Promise = oldPromise;
        });
    });
  });
});


test/request.js
'use strict';

const Request = require('../lib/request'),
  should = require('chai').should(),
  nock = require('nock'),
  HttpsAgent = require('agentkeepalive').HttpsAgent,
  zlib = require('zlib');

describe('Request', () => {
  const self = { };

  const request = new Request('api.bigcommerce.com', { headers: { 'Content-Type': 'application/json' } });

  afterEach(() => nock.cleanAll());

  context('given a missing hostname', () => {
    it('should return an error if hostname is missing', () => {
      /* eslint no-new: off */
      should.Throw(() => {
        new Request();
      }, Error);
    });
  });

  context('given a 429 status code', () => {
    beforeEach(() => {
      self.ordersCall = nock('https://api.bigcommerce.com')
        .post('/orders')
        .reply(429, { }, { 'X-Retry-After': 0.1 })
        .post('/orders')
        .reply(200, { });
    });

    it('should retry the request', () => {
      return request.run('post', '/orders')
        .then(() => self.ordersCall.isDone().should.equal(true));
    });

    context('given a failOnLimitReached option', () => {
      const failRequest = new Request('api.bigcommerce.com', {
        headers: { 'Content-Type': 'application/json' },
        failOnLimitReached: true
      });

      it('should return an error', () => {
        return failRequest.run('post', '/orders')
          .then(() => should.fail('You shall not pass!'))
          .catch(e => {
            e.message.should.match(/rate limit/);
            e.retryAfter.should.equal(0.1);
          });
      });
    });
  });

  context('given a bad request or internal error is returned', () => {
    beforeEach(() => {
      nock('https://api.bigcommerce.com')
        .post('/orders')
        .reply(400, {});
    });

    it('should return an error', () => {
      return request.run('post', '/orders', { })
        .then(() => should.fail('You shall not pass!'))
        .catch(e => e.message.should.match(/Request returned error code/));
    });
  });

  context('if "error" are found in the response JSON', () => {
    beforeEach(() => {
      nock('https://api.bigcommerce.com')
        .post('/orders')
        .reply(200, { error: 'An error has occurred.' });
    });

    it('should return an error', () => {
      return request.run('post', '/orders', { })
        .then(() => should.fail('You shall not pass!'))
        .catch(e => e.message.should.match(/An error has occurred/));
    });
  });

  context('if "errors" are found in the response JSON', () => {
    beforeEach(() => {
      nock('https://api.bigcommerce.com')
        .post('/orders')
        .reply(200, { errors: ['An error has occurred.'] });
    });

    it('should return an error', () => {
      return request.run('post', '/orders', { })
        .then(() => should.fail('You shall not pass!'))
        .catch(e => e.message.should.match(/An error has occurred/));
    });
  });

  context('given a malformed request JSON', () => {
    beforeEach(() => {
      nock('https://api.bigcommerce.com')
        .defaultReplyHeaders({ 'Content-Type': 'application/json' })
        .post('/orders')
        .reply(200, '<malformed>');
    });

    it('should return an error', () => {
      return request.run('post', '/orders', { })
        .then(() => should.fail('You shall not pass!'))
        .catch(e => e.message.should.match(/Unexpected token/));
    });
  });

  context('if json is not returned', () => {
    beforeEach(() => {
      nock('https://api.bigcommerce.com')
        .defaultReplyHeaders({ 'Content-Type': 'application/xml' })
        .post('/orders')
        .reply(200, '<xml></xml>');
      nock('https://api.bigcommerce.com')
        .defaultReplyHeaders({ 'Content-Type': 'application/json' })
        .post('/customers')
        .reply(200, '<html></html>');
    });

    it('should return the raw response', () => {
      return request.run('post', '/orders', { })
        .then(res => res.should.equal('<xml></xml>'));
    });

    it('should attach the response if the JSON cannot be parsed', () => {
      return request.run('post', '/customers', { })
        .catch(err => err.should.have.property('responseBody'));
    });
  });

  context('timeout', () => {
    beforeEach(() => {
      nock('https://api.bigcommerce.com')
        .post('/orders')
        .replyWithError('ECONNRESET');
    });

    it('should return an error', () => {
      return request.run('post', '/orders', { })
        .then(() => should.fail('You shall not pass!'))
        .catch(e => e.message.should.match(/ECONNRESET/));
    });
  });

  it('should attach a keep-alive HTTPS agent', () => {
    nock('https://api.bigcommerce.com')
      .post('/orders')
      .reply(200, { order: true });

    const request = new Request('api.bigcommerce.com', {
      headers: { 'Content-Type': 'application/json' },
      agent: new HttpsAgent({
        maxSockets: 30,
        maxFreeSockets: 30,
        timeout: 60000,
        keepAliveTimeout: 30000
      })
    });

    return request.run('post', '/orders')
      .then(res => res.should.be.a('object'));
  });

  it('should return a JSON object on success', () => {
    nock('https://api.bigcommerce.com')
      .post('/orders')
      .reply(200, { order: true });

    return request.run('post', '/orders')
      .then(res => {
        res.should.be.a('object');
        res.order.should.equal(true);
      });
  });

  it('should accept and parse a GZIP JSON response', () => {
    const data = JSON.stringify({ order: true });
    const buffer = Buffer.from(data);
    const zipped = zlib.gzipSync(buffer);
    nock('https://api.bigcommerce.com')
      .post('/orders')
      .reply(200, zipped, {
        'X-Transfer-Length': String(zipped.length),
        'Content-Length': undefined,
        'Content-Encoding': 'gzip',
        'Content-Type': 'application/json'
      });

    const request = new Request('api.bigcommerce.com', {
      headers: {
        'Content-Type': 'application/json',
        'Accept-Encoding': 'gzip, deflate'
      }
    });

    return request.run('post', '/orders')
      .then(res => {
        should.exist(res);
        res.should.have.property('order', true);
      });
  });
});


lib/bigcommerce.js
'use strict';

/**
 * BigCommerce OAuth2 Authentication and API access
 *
 * @param {Object} config
 * @return null
 *
 * Example Config
 * {
 *   logLevel: 'info',
 *   clientId: 'hjasdfhj09sasd80dsf04dfhg90rsds',
 *   secret: 'odpdf83m40fmxcv0345cvfgh73bdwjc',
 *   callback: 'https://mysite.com/bigcommerce'
 *   accessToken: 'ly8cl3wwcyj12vpechm34fd20oqpnl',
 *   storeHash: 'x62tqn',
 *   responseType: 'json'
 * }
 */

const logger = require('debug')('node-bigcommerce:bigcommerce'),
  crypto = require('crypto'),
  Request = require('./request');

class BigCommerce {
  constructor(config) {
    if (!config) {
      throw new Error(
        'Config missing. The config object is required to make any call to the ' +
        'BigCommerce API'
      );
    }

    this.config = config;
    this.apiVersion = this.config.apiVersion || 'v2';
  }

  verify(signedRequest) {
    if (!signedRequest) {
      throw new Error('The signed request is required to verify the call.');
    }

    const splitRequest = signedRequest.split('.');
    if (splitRequest.length < 2) {
      throw new Error(
        'The signed request will come in two parts seperated by a .(full stop). ' +
        'this signed request contains less than 2 parts.'
      );
    }

    const signature = Buffer.from(splitRequest[1], 'base64').toString('utf8');
    const json = Buffer.from(splitRequest[0], 'base64').toString('utf8');
    const data = JSON.parse(json);

    logger('JSON: ' + json);
    logger('Signature: ' + signature);

    const expected = crypto.createHmac('sha256', this.config.secret)
      .update(json)
      .digest('hex');

    logger('Expected Signature: ' + expected);

    if (
      expected.length !== signature.length ||
      !crypto.timingSafeEqual(Buffer.from(expected, 'utf8'), Buffer.from(signature, 'utf8'))
    ) {
      throw new Error('Signature is invalid');
    }

    logger('Signature is valid');
    return data;
  }

  async authorize(query) {
    if (!query) throw new Error('The URL query paramaters are required.');

    const payload = {
      client_id: this.config.clientId,
      client_secret: this.config.secret,
      redirect_uri: this.config.callback,
      grant_type: 'authorization_code',
      code: query.code,
      scope: query.scope,
      context: query.context
    };

    const request = new Request('login.bigcommerce.com', {
      failOnLimitReached: this.config.failOnLimitReached
    });

    return await request.run('post', '/oauth2/token', payload);
  }

  createAPIRequest() {
    const accept = this.config.responseType === 'xml' ? 'application/xml' : 'application/json';

    return new Request('api.bigcommerce.com', {
      headers: {
        Accept: accept,
        'X-Auth-Client': this.config.clientId,
        'X-Auth-Token': this.config.accessToken
      },
      failOnLimitReached: this.config.failOnLimitReached,
      agent: this.config.agent
    });
  }

  async request(type, path, data) {
    if (!this.config.accessToken || !this.config.storeHash) {
      throw new Error(
        'Get request error: the access token and store hash are required to ' +
        'call the BigCommerce API'
      );
    }

    const extension = this.config.responseType === 'xml' ? '.xml' : '';
    const version = this.apiVersion;

    const request = this.createAPIRequest();

    let fullPath = `/stores/${this.config.storeHash}/${version}`;
    if (version !== 'v3') {
      fullPath += path.replace(/(\?|$)/, extension + '$1');
    } else {
      fullPath += path;
    }

    return await request.run(type, fullPath, data);
  }

  async get(path) {
    return await this.request('get', path);
  }

  async post(path, data) {
    return await this.request('post', path, data);
  }

  async put(path, data) {
    return await this.request('put', path, data);
  }

  async delete(path) {
    return await this.request('delete', path);
  }
}

module.exports = BigCommerce;


lib/request.js
'use strict';

/**
 * To make the requests asynchronously
 */

const logger = require('debug')('node-bigcommerce:request'),
  https = require('https'),
  zlib = require('zlib');

const { version } = require('../package.json');

/**
 * Parse response
 */
function parseResponse(res, body, resolve, reject) {
  try {
    if (!/application\/json/.test(res.headers['content-type']) || body.trim() === '') {
      return resolve(body);
    }

    const json = JSON.parse(body);

    if (json.error || json.errors) {
      const err = new Error(json.error || JSON.stringify(json.errors));

      return reject(err);
    }

    return resolve(json);
  } catch (e) {
    e.responseBody = body;
    return reject(e);
  }
}

class Request {
  constructor(hostname, { headers = { }, failOnLimitReached = false, agent = null } = { }) {
    if (!hostname) throw new Error('The hostname is required to make the call to the server.');

    this.hostname = hostname;
    this.headers = headers;
    this.failOnLimitReached = failOnLimitReached;
    this.agent = agent;
  }

  run(method, path, data) {
    logger(`Requesting Data from: https://${this.hostname}${path} Using the ${method} method`);

    const dataString = JSON.stringify(data);

    const options = {
      path,
      hostname: this.hostname,
      method: method.toUpperCase(),
      port: 443,
      headers: Object.assign({
        'User-Agent': 'node-bigcommerce/' + version,
        'Content-Type': 'application/json',
        'Accept-Encoding': 'gzip, deflate'
      }, this.headers)
    };

    if (this.agent) options.agent = this.agent;

    if (data) {
      options.headers['Content-Length'] = Buffer.from(dataString).length;
    }

    logger('Starting Request, with options.', options);

    return new Promise((resolve, reject) => {
      const req = https.request(options, res => {
        logger('Status Returned: ' + res.statusCode);
        logger('Headers Returned: ' + JSON.stringify(res.headers));

        const contentEncoding = res.headers['content-encoding'];
        const shouldUnzip = ['deflate', 'gzip'].indexOf(contentEncoding) !== -1;
        const encoding = shouldUnzip ? 'binary' : 'utf8';
        let body = '';

        res.setEncoding(encoding);

        // If the API limit has been reached
        if (res.statusCode === 429) {
          const timeToWait = res.headers['x-retry-after'];

          if (this.failOnLimitReached) {
            const err = new Error(`You have reached the rate limit for the BigCommerce API. Please retry in ${timeToWait} seconds.`);
            err.retryAfter = Number(timeToWait);

            return reject(err);
          }

          logger(`You have reached the rate limit for the BigCommerce API, we will retry again in ${timeToWait} seconds.`);

          return setTimeout(() => {
            logger('Restarting request call after suggested time');

            this.run(method, path, data)
              .then(resolve)
              .catch(reject);
          }, timeToWait * 1000);
        }

        res.on('data', chunk => body += chunk);

        res.on('end', () => {
          logger('Request complete');

          if (res.statusCode >= 400 && res.statusCode <= 600) {
            const error = new Error(`Request returned error code: ${res.statusCode} and body: ${body}`);
            error.code = res.statusCode;
            error.responseBody = body;

            return reject(error);
          }

          // Use GZIP decompression if required
          if (shouldUnzip) {
            const unzip = contentEncoding === 'deflate' ? zlib.deflate : zlib.gunzip;
            return unzip(Buffer.from(body, encoding), (err, data) => {
              if (err) {
                return reject(err);
              }
              return parseResponse(res, data.toString('utf8'), resolve, reject);
            });
          }

          return parseResponse(res, body, resolve, reject);
        });
      });

      req.on('error', e => reject(e));

      if (data) {
        logger('Sending Data: ' + dataString);
        req.write(dataString);
      }

      req.end();
    });
  }
}

module.exports = Request;
